effect/shared/src/main/scala/scalaz/effect/KleisliIO.scala

```diff
@@ -0,0 +1,441 @@
+package scalaz.effect
+
+import scalaz.Void
+import scala.{ Boolean, Either, Left, PartialFunction, Right, Unit }
+
+/**
+ * A `KleisliIO[E, A, B]` is an effectful function from `A` to `B`, which might
+ * fail with an `E`.
+ *
+ * This is the moral equivalent of `A => IO[E, B]`, and, indeed, `KleisliIO`
+ * extends this function type, and can be used in the same way.
+ *
+ * The main advantage to using `KleisliIO` is that it provides you a means of
+ * importing an impure function `A => B` into `KleisliIO[E, A, B]`, without
+ * actually wrapping the result of the function in an `IO` value.
+ *
+ * This allows the implementation to aggressively fuse operations on impure
+ * functions, which in turn can result in significantly higher-performance and
+ * far less heap utilization than equivalent approaches modeled with `IO`.
+ *
+ * The implementation allows you to lift functions from `A => IO[E, B]` into a
+ * `KleisliIO[E, A, B]`. Such functions cannot be optimized, but will be handled
+ * correctly and can work in conjunction with optimized (fused) `KleisliIO`.
+ *
+ * Those interested in learning more about modeling effects with `KleisliIO` are
+ * encouraged to read John Hughes paper on the subject: Generalizing Monads to
+ * Arrows (www.cse.chalmers.se/~rjmh/Papers/arrows.pdf). The implementation in
+ * this file contains many of the same combinators as Hughes implementation.
+ *
+ * A word of warning: while even very complex code can be expressed in
+ * `KleisliIO`, there is a point of diminishing return. If you find yourself
+ * using deeply nested tuples to propagate information forward, it may be no
+ * faster than using `IO`.
+ *
+ * Given the following two `KleisliIO`:
+ *
+ * {{{
+ * val readLine = KleisliIO.impureVoid((_ : Unit) => scala.Console.readLine())
+ * val printLine = KleisliIO.impureVoid((line: String) => println(line))
+ * }}}
+ *
+ * Then the following two programs are equivalent:
+ *
+ * {{{
+ * // Program 1
+ * val program1: IO[Void, Unit] =
+ *   for {
+ *     name <- getStrLn
+ *     _    <- putStrLn("Hello, " + name)
+ *   } yield ())
+ *
+ * // Program 2
+ * val program2: IO[Void, Unit] = (readLine >>> KleisliIO.lift("Hello, " + _) >>> printLine)(())
+ * }}}
+ *
+ * Similarly, the following two programs are equivalent:
+ *
+ * {{{
+ * // Program 1
+ * val program1: IO[Void, Unit] =
+ *   for {
+ *     line1 <- getStrLn
+ *     line2 <- getStrLn
+ *     _     <- putStrLn("You wrote: " + line1 + ", " + line2)
+ *   } yield ())
+ *
+ * // Program 2
+ * val program2: IO[Void, Unit] =
+ *   (readLine.zipWith(readLine)("You wrote: " + _ + ", " + _) >>> printLine)(())
+ * }}}
+ *
+ * In both of these examples, the `KleisliIO` program is faster because it is
+ * able to perform fusion of effectful functions.
+ */
+sealed trait KleisliIO[E, A, B] extends (A => IO[E, B]) { self =>
```

Comment:
Extending functions makes me worried

Reply:
Why?

Reply:
Seems like containing a value of this type is just as useful. What's the benefit of extension here?

Reply:
I think the (tiny) benefit is that you can use it wherever `A => IO[E, B]` is required.

Why prefer `(run: A => IO[E, B])`?

Reply:
At the very least, subtyping can interact with partial unification in surprising ways. Consider:

```scala
val foo: KleisliIO[Foo, Bar, Baz] = ...

def bar[F[_], G[_], A](x: F[G[A]]) = x

val baz = foo.map(f)
val qux = bar(foo).map(f)
```

Quiz: Are `baz` and `qux` the same?

Answer: It depends on the definition of `Baz`!

Revised Quiz: What are the semantics of `baz` and `qux` for each of the following definitions of `Baz`?:

```scala
trait Baz
type Baz = Qux[Int]
```

Now seemingly innocuous changes like refactoring `trait Baz` to `type Baz = Qux[Int]` can break code, and with sufficient polymorphism can even cause silent semantic changes rather than type errors.

There may be other problems as well, but I'd have to think about it, and I'd rather just write `.run`

Reply:
@aaronvargo Sounds good. I'll submit a PR. 👍